failed to finalize context because context is zero-initialized

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 20.04
- Installation type:
  - source
- Version or commit hash:
  - e73b613a019e9ff763f68fa3b9bf2300789c1491
- DDS implementation:
  - Fast-RTPS
- Client library (if applicable):
  - rclcpp


#### Steps to reproduce issue

```
ros2 run demo_nodes_cpp talker --ros-args -r chatter:=
```

#### Expected behavior

```
[ERROR] [1601433635.767755175] [rcl]: Failed to parse global arguments
```

#### Actual behavior

```
[ERROR] [1601433635.767755175] [rcl]: Failed to parse global arguments
[ERROR] [1601433635.767803975] [rclcpp]: failed to finalize context: context is zero-initialized, at /tmp/src/ros2/rcl/rcl/src/rcl/context.c:52
```

confirmed as following,

```
# ros2 run demo_nodes_cpp talker --ros-args -r chatter:=
[ERROR] [1601442967.220148940] [rcl]: Failed to parse global arguments

>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'Couldn't parse remap rule: '-r chatter:='. Error: error not set, at /root/ros2_ws/colcon_ws/src/ros2/rcl/rcl/src/rcl/arguments.c:371'

with this new error message:

  'context is zero-initialized, at /root/ros2_ws/colcon_ws/src/ros2/rcl/rcl/src/rcl/context.c:52'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
[ERROR] [1601442967.220200690] [rclcpp]: failed to finalize context: context is zero-initialized, at /root/ros2_ws/colcon_ws/src/ros2/rcl/rcl/src/rcl/context.c:52
terminate called after throwing an instance of 'rclcpp::exceptions::RCLInvalidROSArgsError'
  what():  failed to initialize rcl: error not set

```

--- a/rclcpp/src/rclcpp/context.cpp
+++ b/rclcpp/src/rclcpp/context.cpp
@@ -199,13 +199,17 @@ Context::init(
     throw rclcpp::ContextAlreadyInitialized();
   }
   this->clean_up();
-  rcl_context_.reset(new rcl_context_t, __delete_context);
-  *rcl_context_.get() = rcl_get_zero_initialized_context();
-  rcl_ret_t ret = rcl_init(argc, argv, init_options.get_rcl_init_options(), rcl_context_.get());
+  rcl_context_t * context = new rcl_context_t;
+  if (!context) {
+    throw std::runtime_error("failed to allocate memory for rcl context");
+  }
+  *context = rcl_get_zero_initialized_context();
+  rcl_ret_t ret = rcl_init(argc, argv, init_options.get_rcl_init_options(), context);
   if (RCL_RET_OK != ret) {
-    rcl_context_.reset();
+    delete context;
     rclcpp::exceptions::throw_from_rcl_error(ret, "failed to initialize rcl");
   }
+  rcl_context_.reset(context, __delete_context);
 
   if (init_options.auto_initialize_logging()) {
     logging_mutex_ = get_global_logging_mutex();

